File: Tests/Tests.PagmoSharp/TestProblems/TwoDimensionalConstrainedProblem.cs
using System;
using NUnit.Framework;
using pagmo;

namespace Tests.PagmoSharp.TestProblems
{
    public class TwoDimensionalConstrainedProblem : TestProblemWrapper
    {
        private readonly DoubleVector _lowerBounds = new DoubleVector(new[] { -1.0, -1.0 });
        private readonly DoubleVector _upperBounds = new DoubleVector(new[] { 5.0, 5.0 });

        public override PairOfDoubleVectors get_bounds()
        {
            return new PairOfDoubleVectors(_lowerBounds, _upperBounds);
        }

        /// <inheritdoc />
        public override DoubleVector fitness(DoubleVector arg0)
        {
            double x = arg0[0];
            double y = arg0[1];

            double obj = Math.Pow(x, 2) + y;
            double yEq2Const = y - 2.0; // y == 2
            double xLtEq1Const = x - 1.0; // x == 1
            return new DoubleVector(new[] { obj, yEq2Const, xLtEq1Const });
        }

        public override string get_name()
        {
            return "Simple 2-D x^2+y test problem with constraint y==2.0, x == 1";
        }

        public override thread_safety get_thread_safety()
        {
            return thread_safety.constant;
        }

        /// <inheritdoc />
        public override uint get_nobj()
        {
            return 1;
        }

        /// <inheritdoc />
        public override uint get_nec()
        {
            return 2;
        }

        /// <inheritdoc />
        public override uint get_nic()
        {
            return 0; //TODO: I couldn't get the ant colony optimizer to converge reliably with a x<=1 constraint
        }

        /// <inheritdoc />
        public override double ExpectedOptimalFunctionValue
        {
            get { return 3.0; }
        }

        /// <inheritdoc />
        public override double[] ExpectedOptimalX
        {
            get { return new[] { 1.0, 2.0 }; }
        }

        [Test]
        public void TestBasicEvaluations()
        {
            using var problem = new TwoDimensionalConstrainedProblem();
            var answer = problem.fitness(1.0, 2.0);
            Assert.AreEqual(3.0, answer[0], 2);
            Assert.AreEqual(0.0, answer[1], 2);
            Assert.AreEqual(0.0, answer[2], 2);

            answer = problem.fitness(2.0, 3.0);
            Assert.AreEqual(7.0, answer[0], 2);
            Assert.AreEqual(1.0, answer[1], 2);
            Assert.AreEqual(-1.0, answer[2], 2);

            answer = problem.fitness(0.0, 2.0);
            Assert.AreEqual(2.0, answer[0], 2);
            Assert.AreEqual(0.0, answer[1], 2);
            Assert.AreEqual(1.0, answer[2], 2);

        }
    }
}

File: java/core/src/test/java/io/github/samthegliderpilot/pagmo4j/testproblems/TwoDimensionalConstrainedProblem.java
package io.github.samthegliderpilot.pagmo4j.testproblems;

import io.github.samthegliderpilot.pagmo4j.*;
import io.github.samthegliderpilot.pagmo4j.problems.*;

/**
 * 2-D constrained test problem: minimise x + y subject to x^2 + y^2 <= 1.
 * Optimum at approximately (-0.707, -0.707), f ≈ -1.414.
 */
public class TwoDimensionalConstrainedProblem extends ManagedProblemBase {

    public static final double OPTIMAL_X0 = -0.707;
    public static final double OPTIMAL_X1 = -0.707;
    public static final double OPTIMAL_F  = -1.414;

    @Override
    public DoubleVector fitness(DoubleVector x) {
        double obj = x.get(0) + x.get(1);
        double constraint = x.get(0) * x.get(0) + x.get(1) * x.get(1) - 1.0;
        return vec(obj, constraint);
    }

    @Override
    public PairOfDoubleVectors get_bounds() {
        return bounds(new double[]{-2.0, -2.0}, new double[]{2.0, 2.0});
    }

    @Override
    public long get_nobj() { return 1L; }

    @Override
    public long get_nic() { return 1L; }

    @Override
    public String get_name() { return "2-D Constrained test problem"; }

    @Override
    public ThreadSafety get_thread_safety() { return ThreadSafety.Constant; }
}
